KIPs/kip-247.md

```diff
@@ -0,0 +1,186 @@
+discussions-to: TBU
+status: Draft
+type: Core
+created: 2025-02-18
+---
+
+## Abstract
+
+This KIP introduces Gasless Transaction (GaslessTx) where users can swap whitelisted fungible tokens into KAIA without possessing any KAIA.
+
+## Motivation
+
+For users who have onboarded onto KAIA from another blockchain via a bridge, it is possible that they do not possess any KAIA tokens and hold only fungible tokens from the bridged assets.
+These users will be unable to perform any on-chain actions unless they acquire KAIA through alternative channels.
+This limitation effectively restricts their ability to engage with decentralized applications, execute transactions, or participate in any protocol activities within the KAIA ecosystem.
+
+To enhance the user experience for those, this KIP introduces a concept of gasless transaction (GaslessTx) where transactions with a specific pattern is recognized as the GaslessTx.
+Upon detection of GaslessTx, the block proposer may lend the user KAIA to facilitate the GaslessTx execution.
+A gasless swap from Token to KAIA shall pay back the lent amount.
+
+## Specification
+
+### Definition of GaslessTx
+
+There are two types of GaslessTx: GaslessApproveTx, and GaslessSwapTx.
+
+#### GaslessApproveTx
+
+A GaslessApproveTx must meet all the following conditions to be inserted to `txpool.queue`:
+
+- A1: `GaslessApproveTx.to` is a whitelisted ERC-20 token.
+- A2: `GaslessApproveTx.data` is `approve(spender, amount)`.
+- A3: `spender` is a whitelisted `GaslessSwapRouter`.
+- A4: `amount` is a nonzero.
+
+(Note that above statements can be verified solely through a static validation of a transaction.)
+
+A GaslessApproveTx must meet all the following conditions to be promoted to `txpool.pending`:
+
+- AP1: GaslessApproveTx is followed by a GaslessSwapTx of the same sender and the `GaslessSwapTx` can be promoted to `txpool.pending`.
+
+#### GaslessSwapTx
+
+A `GaslessSwapTx` must meet all the following conditions to be inserted to `txpool.queue`:
+
+- S1: `GaslessSwapTx.to` is a whitelisted `GaslessSwapRouter`.
+- S2: `GaslessSwapTx.data` is `swapForGas(token, amountIn, amountOut, amountRepay)`.
+- S3. `token` is a whitelisted ERC20 token.
+
+(Note that above statements can be verified solely through a static validation of a transaction.)
+
+A `GaslessSwapTx` must meet all the following conditions to be promoted to `txpool.pending`:
+
+- SP1: (If GaslessApproveTx exists in `txpool.queue`) `GaslessApproveTx.to=token`.
+- SP2: (If GaslessApproveTx exists in `txpool.queue`) `GaslessApproveTx.data.amount>=amountIn`.
+- SP3: `amountRepay` is the correct value.
+  - (If GaslessApproveTx exists in `txpool.queue`) `amountRepay = CalcRepayAmount(GaslessApproveTx, GaslessSwapTx)`.
+  - (otherwise) `amountRepay = amountRepay(GaslessSwapTx)`.
+  - `amountRepay(GaslessApproveTx, GaslessSwapTx) = R1 + R2 + R3` and `amountRepay(GaslessSwapTx) = R1 + R3` where
+    - `R1 = LendTx.Fee() = SwapTx.GasPrice() * 21000`
```

Comment:
Consider renaming the amountRepay components to avoid collision with Router conditions. e.g. V1 + V2 + V3.